// n12/dist/index.js
"use strict";
// 🏭 GENERATED WITH generate-main-exports
// ⚠️ Warning: Do not edit by hand, all changes will be lost on next execution!
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizeToSCREAMING_CASE = exports.capitalizeFirstLetter = exports.normalizeToCamelCase = exports.removeDiacritics = exports.searchKeywords = exports.nameToUriPart = exports.parseKeywords = void 0;
var capitalizeFirstLetter_1 = require("./capitalizeFirstLetter");
Object.defineProperty(exports, "capitalizeFirstLetter", { enumerable: true, get: function () { return capitalizeFirstLetter_1.capitalizeFirstLetter; } });
var nameToUriPart_1 = require("./nameToUriPart");
Object.defineProperty(exports, "nameToUriPart", { enumerable: true, get: function () { return nameToUriPart_1.nameToUriPart; } });
var normalizeToCamelCase_1 = require("./normalizeToCamelCase");
Object.defineProperty(exports, "normalizeToCamelCase", { enumerable: true, get: function () { return normalizeToCamelCase_1.normalizeToCamelCase; } });
var normalizeToSCREAMING_CASE_1 = require("./normalizeToSCREAMING_CASE");
Object.defineProperty(exports, "normalizeToSCREAMING_CASE", { enumerable: true, get: function () { return normalizeToSCREAMING_CASE_1.normalizeToSCREAMING_CASE; } });
var parseKeywords_1 = require("./parseKeywords");
Object.defineProperty(exports, "parseKeywords", { enumerable: true, get: function () { return parseKeywords_1.parseKeywords; } });
var removeDiacritics_1 = require("./removeDiacritics");
Object.defineProperty(exports, "removeDiacritics", { enumerable: true, get: function () { return removeDiacritics_1.removeDiacritics; } });
var searchKeywords_1 = require("./searchKeywords");
Object.defineProperty(exports, "searchKeywords", { enumerable: true, get: function () { return searchKeywords_1.searchKeywords; } });
