// wdio-vscode-service/dist/pageobjects/menu/ContextMenu.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { Menu, MenuItem } from '../index.js';
import { PageDecorator, sleep } from '../utils.js';
import { ContextMenu as ContextMenuLocators } from '../../locators/1.73.0.js';
/**
 * Object representing a context menu
 *
 * @category Menu
 */
export let ContextMenu = class ContextMenu extends Menu {
    constructor() {
        super(...arguments);
        /**
         * @private
         */
        this.locatorKey = 'ContextMenu';
    }
    /**
     * Get context menu item by name
     * @param name name of the item to search by
     * @returns Promise resolving to ContextMenuItem object
     */
    async getItem(name) {
        try {
            const items = await this.getItems();
            for (const item of items) {
                if (await item.getLabel() === name) {
                    return item;
                }
            }
            return undefined;
        }
        catch (err) {
            return undefined;
        }
    }
    /**
     * Get all context menu items
     * @returns Promise resolving to array of ContextMenuItem objects
     */
    async getItems() {
        const items = [];
        const elements = await this.itemElement$$;
        for (const element of elements) {
            const classProperty = await element.getAttribute('class');
            if (classProperty.indexOf('disabled') < 0) {
                const item = new ContextMenuItem(this.locatorMap, 
                // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
                element, this);
                await item.wait();
                items.push(item);
            }
        }
        return items;
    }
    /**
     * Close the context menu
     * @returns Promise resolving when the menu is closed
     */
    async close() {
        await browser.keys('Escape');
        await this.elem.waitForDisplayed({ reverse: true });
        // Todo(Christian): maybe handle stale element exception
    }
    /**
     * Wait for the menu to appear and load all its items
     */
    async wait(timeout = 5000) {
        await (await this.elem).waitForDisplayed({ timeout });
        let items = (await this.getItems()).length;
        await browser.waitUntil(async () => {
            const temp = (await this.getItems()).length;
            if (temp === items) {
                return true;
            }
            items = temp;
            return false;
        }, { timeout: 1000 });
        return this;
    }
};
ContextMenu = __decorate([
    PageDecorator(ContextMenuLocators)
], ContextMenu);
/**
 * Object representing an item of a context menu
 *
 * @category Menu
 */
export let ContextMenuItem = class ContextMenuItem extends MenuItem {
    constructor(locators, base, parentMenu) {
        super(locators, base, parentMenu.elem);
        this.parentMenu = parentMenu;
        /**
         * @private
         */
        this.locatorKey = 'ContextMenu';
        this.label = '';
    }
    async select() {
        const nesting = await this.isNesting();
        await this.elem.click();
        await sleep(500);
        if (nesting) {
            const menu = new ContextMenu(this.locatorMap, this.elem);
            await menu.wait();
            return menu;
        }
        return undefined;
    }
    async getLabel() {
        const labelItem = await this.itemLabel$;
        return labelItem.getAttribute(this.locators.itemText);
    }
    async isNesting() {
        try {
            await this.itemNesting$.waitForDisplayed();
            return true;
        }
        catch (err) {
            return false;
        }
    }
};
ContextMenuItem = __decorate([
    PageDecorator(ContextMenuLocators)
], ContextMenuItem);
